src/modules/lua.c

```diff
@@ -664,6 +665,88 @@ mtev_console_show_lua(mtev_console_closure_t ncct,
   return 0;
 }
 
+static int
+mtev_lua_xcall_reporter(eventer_t e, int mask, void *closure,
+                               struct timeval *now) {
+  (void)e;
+  (void)mask;
+  (void)now;
+  struct lua_reporter *reporter = closure;
+  mtev_hash_iter zero = MTEV_HASH_ITER_ZERO, iter;
+  pthread_t me = pthread_self();
+  mtev_json_object *states = NULL;
+
+  mtevAssert(reporter->approach == LUA_REPORT_JSON);
+
+  pthread_mutex_lock(&reporter->lock);
+  states = reporter->root;
+  memcpy(&iter, &zero, sizeof(zero));
+  pthread_mutex_lock(&mtev_lua_states_lock);
+  while(mtev_hash_adv(&mtev_lua_states, &iter)) {
+    lua_State *L = *((lua_State **) iter.key.ptr);
+    lua_module_closure_t *lmc = iter.value.ptr;
+    pthread_t tgt = lmc->owner;
+    char state_str[32];
+    mtev_json_object *out;
+    if(!pthread_equal(me, tgt)) continue;
+    snprintf(state_str, sizeof(state_str), "%p", (void*)L);
+    lua_getglobal(L, "xcall");
+    if(lua_isnil(L, -1)) {
+      out = MJ_STR("<xcall function not found>");
+    }
+    else {
+      mtev_lua_hash_to_table(L, reporter->args);
+      /*  Invoke xcall() */
+      lua_call(L, 1, 1);
+      /* Convert results to json via mtev.tojson(...):unwrap() */
+      lua_getglobal(L, "mtev");
+      lua_getfield(L, -1, "tojson");
+      lua_remove(L, -2);
+      lua_insert(L, -2);
+      lua_call(L, 1, 1);
+      lua_getfield(L, -1, "unwrap");
+      lua_insert(L, -2);
+      lua_call(L, 1, 1);
+      mtev_json_object **udata = luaL_checkudata(L, -1, "mtev.json_object");
+      out = (*udata);
```

Comment:
Not sure about the ownership of that json object here. Do we need to explicitly free/unref it after use?

- The `:uwrap()` function calls `mtev_json_get()` so the object survives the lua function's return.
- mtev_rest_show_lua_complete calls `mtev_json_put()` on the root object, which should (?) unref this value here as well.